### DISCUSSION TITLE
How can python code be used to generate rows in a tabularray?

### DISCUSSION THREAD

**Discussion by kiryph** [lvjr/tabularray#289] [Q&A]
[pythontex](https://ctan.org/pkg/pythontex) describes two options to use python code within a `tabular` environment:
> - The `tabular` environment can conflict with PythonTeX under some circumstances, due to how `tabular` functions. Among other things, printing within a `tabular` environment can cause errors, because printing involves bringing in external content via `\InputIfFileExists`, but that macro is not _expandable_. There are a few different ways to work around the limitations of `tabular`. 
>   - Put the printed content in a macro definition, and use the macro in `tabular`. You will have to create a dummy version of the macro, to avoid errors before the macro is defined by PythonTeX. An example is given below. The` \global\def` is needed so that the macro is defined outside of the `pycode` environment. 
>   ```tex
>    \let\row\relax
>    \begin{pycode}
>    print("\\global\\def\\row{a & b & c & d & e \\\\}")
>    \end{pycode}
>  
>    \begin{tabular}{|c|c|c|c|c|}
>    \row
>    \end{tabular} 
>   ```
>    - Use `\py`. The end-of-row `\\` must be outside of the command. Example:
>   ```tex
>    \begin{tabular}{|c|c|c|c|c|}
>    \py{"a & b & c & d & e"} \\
>    \end{tabular} 
>   ```

Unfortunately, both do not work with tabularray.

**Is there anything else to consider?**

IMHO using an embedded scripting language is in particular useful for tables.
Other options might be luatex or perltex. However, I import a [graphml](http://graphml.graphdrawing.org/) file with [networkx](https://networkx.org/) which makes python the right choice of scripting language.
The end goal is a similar matrix as following one 
![Z2^4;_Lattice_of_subgroups_Hasse_diagram_adjacency_matrix svg](https://user-images.githubusercontent.com/371756/185058001-9a8b0e3c-a715-412e-856f-d22c7a054fee.png)

https://commons.wikimedia.org/wiki/File:Z2%5E4;_Lattice_of_subgroups_Hasse_diagram_adjacency_matrix.svg

### Minimal Failing Example using Pyluatex
(single lualatex run to generate pdf)
```tex
%! TeX program = lualatex
\documentclass{article}
\usepackage{pyluatex}
\usepackage{tabularray}

\begin{document}
% needs to be after document
\let\row\relax
\begin{python}
print("\\global\\def\\row{a & b & c & d & e \\\\}")
\end{python}

\begin{tblr}{|c|c|c|c|c|}
a & b & c & d & e \\
\row
\end{tblr}

\end{document}
```

```
❯ lualatex --output-format=pdf --shell-escape pythontex-tabularray.tex
This is LuaHBTeX, Version 1.15.0 (TeX Live 2022)
 system commands enabled.
(./pythontex-tabularray.tex
LaTeX2e <2021-11-15> patch level 1
 L3 programming layer <2022-02-24>
(/usr/local/texlive/2022/texmf-dist/tex/latex/base/article.cls
Document Class: article 2021/10/04 v1.4n Standard LaTeX document class
(/usr/local/texlive/2022/texmf-dist/tex/latex/base/size10.clo))
(/usr/local/texlive/2022/texmf-dist/tex/lualatex/pyluatex/pyluatex.sty
(/usr/local/texlive/2022/texmf-dist/tex/latex/l3kernel/expl3.sty
(/usr/local/texlive/2022/texmf-dist/tex/latex/l3backend/l3backend-luatex.def))
(/usr/local/texlive/2022/texmf-dist/tex/latex/kvoptions/kvoptions.sty
(/usr/local/texlive/2022/texmf-dist/tex/latex/graphics/keyval.sty)
(/usr/local/texlive/2022/texmf-dist/tex/generic/ltxcmds/ltxcmds.sty)
(/usr/local/texlive/2022/texmf-dist/tex/generic/kvsetkeys/kvsetkeys.sty))
(/usr/local/texlive/2022/texmf-dist/tex/latex/base/atveryend-ltx.sty))
(/usr/local/texlive/2022/texmf-dist/tex/latex/tabularray/tabularray.sty)
(./pythontex-tabularray.aux)
(/usr/local/texlive/2022/texmf-dist/tex/latex/base/ts1cmr.fd)
! Misplaced alignment tab character &.
\row ->a &
           b & c & d & e \\
l.17 \end
       {tblr}
?
```

**Comment by lvjr** [comment]
Sorry, I am not able to visit wikimedia website.

You need to use outer key `expand` with `tblr` environment:
```latex
 \begin{tblr}[expand=\row]{|c|c|c|c|c|}
 \row
 \end{tblr} 
```
Also you can use `functional` library of `tabularray` package to build table rows. See the manual and this example https://tex.stackexchange.com/questions/645074/generate-tabularray-and-compute-sub-result-for-each-row-using-functional-library

**Reply by kiryph** [reply]
> Sorry, I am not able to visit wikimedia website.

I have inserted a screenshot.
